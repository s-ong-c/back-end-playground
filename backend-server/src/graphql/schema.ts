import merge from 'lodash/merge';
import * as post from './post';
import * as user from './user';
import * as comment from './comment';
import * as series from './series';
import { gql, IResolvers, makeExecutableSchema } from 'apollo-server-koa';
import DateScalar from './scalars/dateScalar';

const typeDef = gql`
  scalar JSON
  scalar Date
  type Query {
    _version: String
  }
  type Mutation {
    _empty: String
  }
`;

const resolvers: IResolvers = {
  Query: {
    _version: () => '1.0'
  },
  Mutation: {},
  Date: DateScalar
};

const schema = makeExecutableSchema({
  typeDefs: [typeDef, user.typeDef, post.typeDef, comment.typeDef, series.typeDef],
  resolvers: merge(resolvers, user.resolvers, post.resolvers, comment.resolvers, series.resolvers)
});

export default schema;
