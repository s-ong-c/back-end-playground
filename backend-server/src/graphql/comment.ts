import { gql, IResolvers } from 'apollo-server-koa';
import Comment from '../entity/Comment';
import { getRepository } from 'typeorm';
import User, { userLoader } from '../entity/User';

export const typeDef = gql`
  type Comment {
    id: ID!
    text: String
    likes: Int
    level: Int
    has_replies: Boolean
    deleted: Boolean
    user: User
    replies: [Comment]
    created_at: Date
  }
  extend type Query {
    subcomments(comment_id: ID): [Comment]
  }
`;

export const resolvers: IResolvers = {
  Comment: {
    text: (parent: Comment) => {
      if (parent.deleted) {
        return null;
      }
      return parent.text;
    },
    user: (parent: Comment) => {
      if (parent.deleted) return null;
      if (parent.user) return parent.user;
      const user = userLoader.load(parent.fk_user_id);
      return user;
    }
  },
  Query: {
    subcomments: async (parent: any, { comment_id }) => {
      const comments = await getRepository(Comment).find({
        where: {
          reply_to: comment_id
        },
        order: {
          created_at: 'ASC'
        }
      });
      return comments;
    }
  }
};