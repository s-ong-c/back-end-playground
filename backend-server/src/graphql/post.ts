import { ApolloContext } from './../app';
import { gql, IResolvers, ApolloError, AuthenticationError } from 'apollo-server-koa';
import Post from '../entity/Post';
import { getRepository, getManager, getConnectionManager } from 'typeorm';
import PostScore from '../entity/PostScore';
import { normalize } from '../lib/utils';
import removeMd from 'remove-markdown';
import PostsTags from '../entity/PostsTags';
import Tag from '../entity/Tag';

export const typeDef = gql`
  type Post {
    id: ID!
    title: String
    body: String
    thumbnail: String
    is_markdown: Boolean
    is_temp: Boolean
    user: User
    url_slug: String
    likes: Int
    meta: JSON
    views: Int
    is_private: Boolean
    released_at: Date
    created_at: Date
    updated_at: Date
    short_description: String
    comments: [Comment]
    tags: [String]
    comments_count: Int
  }
  extend type Query {
    post(id: ID, username: String, url_slug: String): Post
    posts(cursor: ID, limit: Int, username: String): [Post]
    trendingPosts(offset: Int, limit: Int): [Post]
  }
  extend type Mutation {
    writePost(
      title: String
      body: String
      tags: [String]
      is_markdown: Boolean
      is_temp: Boolean
      url_slug: String
      thumbnail: String
      meta: JSON
      is_private: Boolean
    ): Post
  }
`;

type WritePostArgs = {
  title: string;
  body: string;
  tags: string[];
  is_markdown: boolean;
  is_temp: boolean;
  url_slug: string;
  thumbnail: string | null;
  meta: any;
};

export const resolvers: IResolvers<any, ApolloContext> = {
  Post: {
    user: (parent: Post, _: any, { loaders }) => {
      if (!parent.user) {
        return loaders.user.load(parent.fk_user_id);
      }
      // TODO: fetch user if null
      return parent.user;
    },
    short_description: (parent: Post) => {
      if (parent.meta.short_description) {
        return parent.meta.short_description;
      }

      const removed = removeMd(
        parent.body
          .replace(/```([\s\S]*?)```/g, '')
          .replace(/~~~([\s\S]*?)~~~/g, '')
          .slice(0, 500)
      );
      return removed.slice(0, 200) + (removed.length > 200 ? '...' : '');
    },
    comments: (parent: Post, _: any, { loaders }) => {
      if (parent.comments) return parent.comments;
      return loaders.comments.load(parent.id);
    },
    comments_count: async (parent: Post, _: any, { loaders }) => {
      if (parent.comments) return parent.comments.length;
      const comments = await loaders.comments.load(parent.id);
      return comments.length;
    },
    tags: async (parent: Post, _: any, { loaders }) => {
      const tags = await loaders.tags.load(parent.id);
      return tags.map(tag => tag.name);
    }
  },
  Query: {
    post: async (parent: any, { id, username, url_slug }: any, ctx: any) => {
      try {
        if (id) {
          const post = await getManager()
            .createQueryBuilder(Post, 'post')
            .leftJoinAndSelect('post.user', 'user')
            .leftJoinAndSelect('post.comments', 'comment')
            .where('post.id = :id', { id })
            .orderBy({
              'comment.created_at': 'ASC'
            })
            .getOne();

          return post;
        }
        const post = await getManager()
          .createQueryBuilder(Post, 'post')
          .leftJoinAndSelect('post.user', 'user')
          .leftJoinAndSelect('post.comments', 'comment')
          .where('user.username = :username AND url_slug = :url_slug', { username, url_slug })
          .orderBy({
            'comment.created_at': 'ASC'
          })
          .getOne();
        if (!post) {
          console.log('fuck');
          return null;
        }
        if ((post.is_temp || post.is_private === true) && post.fk_user_id !== ctx.user_id) {
          return null;
        }

        return post;
      } catch (e) {
        console.log(e);
      }
    },
    posts: async (parent: any, { cursor, limit = 20, username }: any, context: any) => {
      const query = getManager()
        .createQueryBuilder(Post, 'post')
        .limit(limit)
        .orderBy('post.released_at', 'DESC')
        .addOrderBy('post.id', 'DESC')
        .leftJoinAndSelect('post.user', 'user')
        .where('post.is_temp = false AND is_private = false');

      if (username) {
        query.andWhere('user.username = :username', { username });
      }

      if (cursor) {
        const post = await getRepository(Post).findOne({
          id: cursor
        });
        if (!post) {
          throw new ApolloError('invalid cursor');
        }
        query.andWhere('post.released_at < :date', {
          date: post.released_at,
          id: post.id
        });
        query.orWhere('post.released_at = :date AND post.id < :id', {
          date: post.released_at,
          id: post.id
        });
      }

      if (context.user_id) {
        query.orWhere('post.is_private = true and post.fk_user_id = :id', {
          id: context.user_id
        });
      }
      const posts = await query.getMany();
      return posts;
    },
    trendingPosts: async (parent: any, { offset = 0, limit = 20 }) => {
      const query = getRepository(PostScore)
        .createQueryBuilder()
        .select('fk_post_id')
        .addSelect('SUM(score)', 'score')
        .where('created_at > now()::DATE - 14 AND fk_post_id IS NOT NULL')
        .groupBy('fk_post_id')
        .orderBy('score', 'DESC')
        .addOrderBy('fk_post_id', 'DESC')
        .limit(limit);

      if (offset) {
        query.offset(offset);
      }

      const rows = (await query.getRawMany()) as { fk_post_id: string; score: number }[];
      const ids = rows.map(row => row.fk_post_id);
      const posts = await getRepository(Post).findByIds(ids);
      const normalized = normalize(posts);
      const ordered = ids.map(id => normalized[id]);

      return ordered;
    }
  },
  Mutation: {
    writePost: async (parent: any, args, ctx) => {
      if (!ctx.user_id) {
        throw new AuthenticationError('Not Logged In');
      }
      const post = new Post();
      const data = args as WritePostArgs;
      post.fk_user_id = ctx.user_id;
      post.title = data.title;
      post.body = data.body;
      post.is_temp = data.is_temp;
      post.is_markdown = data.is_markdown;
      post.meta = data.meta;
      post.thumbnail = data.thumbnail;
      post.url_slug = data.url_slug;

      const postRepo = getRepository(Post);
      const tagsData = await Promise.all(data.tags.map(Tag.findOrCreate));
      await postRepo.save(post);
      PostsTags.syncPostTags(post.id, tagsData);

      post.tags = tagsData;
      return post;
    }
  }
};
