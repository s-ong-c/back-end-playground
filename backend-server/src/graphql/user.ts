import { gql, IResolvers, AuthenticationError } from 'apollo-server-koa';
import User, { userLoader } from '../entity/User';
import { getManager, getRepository } from 'typeorm';
import UserProfile, { userProfileLoader } from '../entity/UserProfile';
import { seriesListLoader } from '../entity/Series';

export const typeDef = gql`
  type User {
    id: ID!
    username: String
    email: String
    created_at: Date
    updated_at: Date
    is_certified: Boolean
    profile: UserProfile
    series_list: [Series]
  }
  type UserProfile {
    id: ID!
    display_name: String
    short_bio: String
    thumbnail: String
    created_at: Date
    updated_at: Date
    about: String
    profile_links: JSON
  }
  extend type Query {
    user(id: ID, username: String): User
    auth: User
  }
`;

export const resolvers: IResolvers = {
  User: {
    profile: async (parent: User) => {
      return userProfileLoader.load(parent.id);
    },
    email: (parent: User, _: any, context: any) => {
      if (context.user_id !== parent.id) {
        throw new AuthenticationError('No permission to read email address');
      }
      return parent.email;
    },
    series_list: (parent: User) => {
      return seriesListLoader.load(parent.id);
    }
  },
  Query: {
    user: async (parent: any, { id, username }: any) => {
      const repo = getRepository(User);
      try {
        if (username) {
          const user = repo.findOne({
            where: {
              username
            }
          });
          return user;
        }
        const user = repo.findOne({
          id
        });
        return user;
      } catch (e) {
        console.log(e);
      }
    },
    auth: async (parent: any, params: any, ctx: any) => {
      if (!ctx.user_id) return null;
      return userLoader.load(ctx.user_id);
    }
  }
};
