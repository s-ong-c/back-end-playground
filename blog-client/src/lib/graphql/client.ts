import ApolloClient from 'apollo-boost';
import { InMemoryCache } from 'apollo-cache-inmemory';
const client = new ApolloClient({
    uri: '/graphql',
    cache: new InMemoryCache().restore((window as any).__APOLLO_STATE__),
});

export default client;