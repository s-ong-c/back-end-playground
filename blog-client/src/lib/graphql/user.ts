import gql from 'graphql-tag';

export type UserProfile = {
  id: string;
  display_name: string;
  short_bio: string;
  thumbnail: string | null;
  about: string;
  profile_links: any;
};
export type SongcConfig = {
  id: string;
  title: string;
};
export type User = {
  id: string;
  username: string;
  email: string | null;
  is_certified: boolean;
  profile: UserProfile;
  songcConfig: SongcConfig | null;
};

export const GET_CURRENT_USER = gql`
  query CurrentUser {
    auth {
      id
      username
      profile {
        thumbnail
        display_name
      }
    }
  }
`;

export const GET_SONGC_CONFIG = gql`
  query GetSongcConfig($username: String) {
    songc_config(username: $username) {
      title
      logo_image
    }
  }
`;

export type CurrentUser = {
  id: string;
  username: string;
  profile: {
    thumbnail: string | null;
    display_name: string;
  };
};
