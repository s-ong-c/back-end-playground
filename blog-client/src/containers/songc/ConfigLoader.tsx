import * as React from 'react';
import { Query, QueryResult } from 'react-apollo';
import { GET_SONGC_CONFIG, SongcConfig } from '../../lib/graphql/user';

export interface ConfigLoaderProps {
  username: string;
}

const ConfigLoader: React.FC<ConfigLoaderProps> = ({ username }) => {
  return (
    <Query query={GET_SONGC_CONFIG} variables={{ username }}>
      {({
        loading,
        data,
        error,
      }: QueryResult<{ songc_config: SongcConfig }>) => {
        if (error) {
          console.log(error);
        }
        if (error || loading) return null;
        if (!data) return null;
        return <div>{data.songc_config.title}</div>;
      }}
    </Query>
  );
};

export default ConfigLoader;
