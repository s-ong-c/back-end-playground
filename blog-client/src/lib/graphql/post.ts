import gql from 'graphql-tag';
import { User } from './user';

export type Post = {
  id: string;
  title: string;
  body: string;
  thumbnail: string;
  is_markdown: boolean;
  is_temp: boolean;
  user: any;
  url_slug: string;
  likes: number;
  meta: any;
  views: number;
  is_private: boolean;
  released_at: string;
  created_at: string;
  updated_at: string;
  short_description: string;
  comments: [any];
  tags: [string];
  comments_count: number;
};

// Post Type for PostList
export type PartialPost = {
  id: string;
  title: string;
  short_description: string;
  thumbnail: string;
  user: User;
  url_slug: string;
  is_private: boolean;
  released_at: string;
  tags: [string];
  comments_count: number;
};

export const GET_POST_LIST = gql`
  query Post($cursor: ID) {
    posts(cursor: $cursor) {
      id
      title
      short_description
      thumbnail
      user {
        id
        username
        profile {
          thumbnail
        }
      }
      url_slug
      released_at
      comments_count
      tags
      is_private
    }
  }
`;

export const WRITE_POST = gql`
  mutation WritePost(
    $title: String
    $body: String
    $tags: [String]
    $is_markdown: Boolean
    $is_temp: Boolean
    $url_slug: String
    $thumbnail: String
    $meta: JSON
  ) {
    writePost(
      title: $title
      body: $body
      tags: $tags
      is_markdown: $is_markdown
      is_temp: $is_temp
      url_slug: $url_slug
      thumbnail: $thumbnail
      meta: $meta
    ) {
      id
      title
      body
      thumbnail
      is_markdown
      is_temp
      user {
        id
        username
        profile {
          thumbnail
        }
      }
      url_slug
      meta
    }
  }
`;
