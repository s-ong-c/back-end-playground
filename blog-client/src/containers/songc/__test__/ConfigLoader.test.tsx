import * as React from 'react';
import { render, waitForDomChange } from 'react-testing-library';
import ConfigLoader, { ConfigLoaderProps } from '../ConfigLoader';
import { MockedProvider } from 'react-apollo/test-utils';
import { GET_SONGC_CONFIG } from '../../../lib/graphql/user';

describe('ConfigLoader', () => {
  const setup = (props: Partial<ConfigLoaderProps> = {}) => {
    const initialProps: ConfigLoaderProps = { username: 'songc' };
    const mocks = [
      {
        request: {
          query: GET_SONGC_CONFIG,
          variables: {
            username: 'songc',
          },
        },
        result: {
          data: {
            songc_config: { title: 'SONGC', logo_image: null },
          },
        },
      },
    ];
    const utils = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <ConfigLoader {...initialProps} {...props} />
      </MockedProvider>,
    );
    return {
      ...utils,
    };
  };
  it('renders property', () => {
    setup();
  });
  it('matches snapshot', async () => {
    const { container } = setup();
    await waitForDomChange({ container });
    expect(container).toMatchSnapshot();
  });
});
