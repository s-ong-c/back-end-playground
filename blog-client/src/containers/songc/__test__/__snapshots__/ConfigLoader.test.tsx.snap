// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigLoader matches snapshot 1`] = `
<div>
  <div>
    SONGC
  </div>
</div>
`;
